MinPlatformPkg/Test/Library/TestPointCheckLib/StandaloneMmCheckPaging.c

```diff
@@ -0,0 +1,419 @@
+
+  //
+  // Below logic is to check 2M/4K page to make sure we donot waist memory.
+  //
+  while (BaseAddress < BaseAddressEnd) {
+    PageEntry = GetPageTableEntry (BaseAddress, &PageAttribute);
+    ASSERT (PageAttribute != PageNone);
+
+    if (IsOutsideSmram) {
+      if ((PageEntry != NULL) && ((*PageEntry & IA32_PG_P) != 0)) {
+        DEBUG ((DEBUG_ERROR, "OutsideSmram present - 0x%x\n", BaseAddress));
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+
+    if (IsCode) {
+      if ((PageEntry == NULL) || ((*PageEntry & IA32_PG_P) == 0)) {
+        DEBUG ((DEBUG_ERROR, "Code Page not exist - 0x%x\n", BaseAddress));
+        return EFI_INVALID_PARAMETER;
+      } else if ((*PageEntry & IA32_PG_RW) != 0) {
+        //
+        // Check if it is StandaloneMM SaveState
+        //
+        if (!IsStandaloneMmSaveState (BaseAddress)) {
+          DEBUG ((DEBUG_ERROR, "Code Page read write - 0x%x\n", BaseAddress));
+          return EFI_INVALID_PARAMETER;
+        }
+      }
+    } else {
+      if ((PageEntry == NULL) || ((*PageEntry & IA32_PG_P) == 0)) {
+        // Pass
+      } else if ((*PageEntry & IA32_PG_NX) == 0) {
+        DEBUG ((DEBUG_ERROR, "Data Page executable - 0x%x\n", BaseAddress));
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+
+    PageEntryLength = PageAttributeToLength (PageAttribute);
+    PageEntryLength = (UINTN)((BaseAddress & ~(PageEntryLength - 1)) + PageEntryLength - BaseAddress);
+
+    //
+    // move to next
+    //
+    BaseAddress += PageEntryLength;
+    Length      -= PageEntryLength;
+  }
+
+  return RETURN_SUCCESS;
+}
+
+EFI_STATUS
+TestPointCheckPagingWithMemoryAttributesTable (
+  IN EFI_MEMORY_ATTRIBUTES_TABLE  *MemoryAttributesTable
+  )
+{
+  UINTN                  Index;
+  EFI_MEMORY_DESCRIPTOR  *Entry;
+  EFI_STATUS             Status;
+  EFI_STATUS             ReturnStatus;
+
+  ReturnStatus = EFI_SUCCESS;
+  Entry        = (EFI_MEMORY_DESCRIPTOR *)(MemoryAttributesTable + 1);
+  for (Index = 0; Index < MemoryAttributesTable->NumberOfEntries; Index++) {
+    DEBUG ((DEBUG_INFO, "StandaloneMmMemoryAttribute Checking 0x%lx - 0x%x\n", Entry->PhysicalStart, EFI_PAGES_TO_SIZE ((UINTN)Entry->NumberOfPages)));
+    Status = TestPointCheckPageTable (
+               Entry->PhysicalStart,
+               EFI_PAGES_TO_SIZE ((UINTN)Entry->NumberOfPages),
+               ((Entry->Attribute & EFI_MEMORY_RO) == 0) ? FALSE : TRUE,
+               FALSE
+               );
+    if (EFI_ERROR (Status)) {
+      ReturnStatus = Status;
+    }
+
+    Entry = NEXT_MEMORY_DESCRIPTOR (Entry, MemoryAttributesTable->DescriptorSize);
+  }
+
+  return ReturnStatus;
+}
+
+EFI_STATUS
+TestPointCheckStandaloneMmPaging (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+
+  // VOID        *MemoryAttributesTable;
+
+  DEBUG ((DEBUG_INFO, "==== TestPointCheckStandaloneMmPaging - Enter\n"));
+
+  /*Status = StandaloneMmGetSystemConfigurationTable(&gEdkiiPiSmmMemoryAttributesTableGuid, (VOID**)&MemoryAttributesTable);
```

Comment:
This seems to be the only substantial change for this file compared to the original smm version? I think we can abstract this function as `TestPointCheckMmPaging` and the standalone mm instance just return unsupported result.

Reply:
Updated.